prov/util/src/util_coll.c

```diff
@@ -764,6 +754,29 @@ static struct fi_ops util_coll_fi_ops = {
 	.ops_open = fi_no_ops_open,
 };
 
+int ofi_av_set_addr(struct fid_av_set *set, fi_addr_t *coll_addr)
+{
+	struct util_av_set *av_set;
+	struct util_coll_mc *coll_mc;
+
+	av_set = container_of(set, struct util_av_set, av_set_fid);
+
+	coll_mc = calloc(1, sizeof(*coll_mc));
+	if (!coll_mc)
+		return -FI_ENOMEM;
+
+	coll_mc->mc_fid.fid.fclass = FI_CLASS_MC;
+	coll_mc->mc_fid.fid.context = NULL;
+	coll_mc->mc_fid.fid.ops = &util_coll_fi_ops;
+	coll_mc->mc_fid.fi_addr = (uintptr_t) coll_mc;
+	coll_mc->av_set = av_set;
+	coll_mc->ep = NULL;
+
+	*coll_addr = (uintptr_t)coll_mc;
+
+	return FI_SUCCESS;
+}
```

Comment:
I have a couple requests while you're in this part of the code.  Can you either delete the util_coll_mc_alloc() routine, or move the above functionality there?  As it is now, it does nothing but wrap calloc.

Also, replace the duplicate allocation code (for new_coll_mc) in fi_join_multicast with whatever new function (even calling ofi_av_set_addr) allocates the structure.  Likewise, it looks like there's duplicate allocation code in util_coll_av_init().

Basically, there should only be 1 place where we allocate struct util_coll_mc.

Reply:
fixed, please review.
Also add another commit to add coll_mc to av_set struct.